test: test for syft base CPU image

khoaguin opened the pull request:
Add a tox test to test Base Image for Syft in CI (Issue: https://github.com/OpenMined/Heartbeat/issues/869)

Also build and publish `syft-base-cpu` to DockerHub (Issue: https://github.com/OpenMined/Heartbeat/issues/870)

kiendang committed:
Enable apk cache dockerfile

kiendang committed:
Improve pip cache

More precise cache location

--mount=type=cache doesn't support the rw option anymore. It's readwrite
by default.

kiendang committed:
Separate cache for different users

kiendang committed:
Simplify COPY

khoaguin committed:
test: on creating base cpu test in `tox.ini`

khoaguin committed:
test: add a build step for `syft_base_cpu.dockerfile` to tox.ini.
- create a step to build and push the image built from `syft_base_cpu.dockerfile` to DockerHub

khoaguin committed:
ci(pr-tests-stack): run tox test for building syft cpu base image to `pr-tests-stack.yml`

kiendang commented on .github/workflows/cd-syft.yml:
```diff
@@ -129,6 +129,18 @@ jobs:
           cache-from: type=registry,ref=openmined/grid-backend:cache-${{ steps.release_metadata.outputs.short_release_platform }}
           cache-to: type=registry,ref=openmined/grid-backend:cache-${{ steps.release_metadata.outputs.short_release_platform }},mode=max
 
+      - name: Build and push `syft-base-cpu` image to DockerHub
+        id: syft-base-cpu-build
+        uses: docker/build-push-action@v5
+        with:
+          context: ./packages
+          file: ./packages/grid/backend/syft_base_cpu.dockerfile
+          platforms: ${{ steps.release_metadata.outputs.release_platform }}
+          target: backend
+          outputs: type=image,name=openmined/syft-base-cpu,push-by-digest=true,name-canonical=true,push=true
+          cache-from: type=registry,ref=openmined/syft-base-cpu:cache-${{ steps.release_metadata.outputs.short_release_platform }}
+          cache-to: type=registry,ref=openmined/syft-base-cpu:cache-${{ steps.release_metadata.outputs.short_release_platform }},mode=max
+
```
👍

kiendang commented on .github/workflows/pr-tests-stack.yml:
```diff
@@ -169,6 +169,7 @@ jobs:
           PYTEST_MODULES: "${{ matrix.pytest-modules }}"
         run: |
           tox -e stack.test.integration
+          tox -e backend.test.basecpu
```
We should create a new job instead of adding to the `pr-tests-stack` job here.

khoaguin committed:
ci(pr-tests-stack): separating running syft backend base image
building test from integration test

khoaguin committed:
Merge branch 'dev' into test/syft_base_cpu

khoaguin committed:
test(syft_base_cpu): add tox tests when using custom `PIP_PACKAGES`
and `CUSTOM_CMD`. Add a new job in `pr-test-stack.yml` for this test
to run on CI

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>
Co-authored-by: Kien Dang <mail@kien.ai>

shubham3121 committed:
fix run argument for custom commands in syft base dockerfile

kiendang commented on tox.ini:
```diff
@@ -1041,3 +1042,46 @@ commands =
+    bash -c 'docker build -f ./grid/backend/syft_base_cpu.dockerfile . -t cpu-worker:latest'
+    bash -c 'docker rmi cpu-worker:latest'
+
+    bash -c 'docker rm $(docker ps -aq) --force || true'
+    bash -c 'docker volume prune -f || true'
+    bash -c 'docker buildx use default || true'
+    bash -c 'docker build -f grid/backend/syft_base_cpu.dockerfile . -t cpu-worker:llama  --build-arg PIP_PACKAGES="llama-index" --build-arg SYSTEM_PACKAGES="htop"'
+    bash -c 'docker run --rm cpu-worker:llama pip list | grep llama-index'
+    bash -c 'docker run --rm cpu-worker:llama apk -e info htop'
+    bash -c 'docker rmi cpu-worker:llama'
+
+    bash -c 'docker rm $(docker ps -aq) --force || true'
+    bash -c 'docker volume prune -f || true'
+    bash -c 'docker buildx use default || true'
+    bash -c 'docker build -f grid/backend/syft_base_cpu.dockerfile . -t cpu-worker:opendp  --build-arg PIP_PACKAGES="llama-index opendp" --build-arg SYSTEM_PACKAGES="curl wget htop"'
+    bash -c 'docker run --rm cpu-worker:opendp pip list | grep llama-index'
+    bash -c 'docker run --rm cpu-worker:opendp pip list | grep opendp'
+    bash -c 'docker run --rm cpu-worker:opendp apk -e info htop'
+    bash -c 'docker run --rm cpu-worker:opendp apk -e info curl'
+    bash -c 'docker run --rm cpu-worker:opendp apk -e info wget'
+    bash -c 'docker rmi cpu-worker:opendp'
```
Are these duplicated? Do we need L1060-L1066?

shubham3121 replied:
Yep, we can only keep the latter (from line 1068).

kiendang replied:
alr solved.

kiendang committed:
Remove unnecessary install

Can just make without install here

kiendang committed:
Allow customizing PIP_PACKAGES and SYSTEM_PACKAGES in tox task

kiendang committed:
Noop cleanup

kiendang commented:
Sometimes the `apk add` in the backend build step fails, but passes when rerun. I reopened #8273 temporarily to investigate further.

kiendang committed:
Add concurrent write lock to apk cache

kiendang commented:
Yup adding a concurrent write lock [`sharing=locked`](https://docs.docker.com/engine/reference/builder/#run---mounttypecache) seems to solve it.

kiendang committed:
Correct rootless comments

kiendang committed:
Use UID through ARG instead of hardcoded

kiendang committed:
Add missing file filter step

Otherwise the job won't run

kiendang committed:
Add missing checkout step

kiendang committed:
builder only needs setting once

kiendang committed:
Add concurrent write lock to pip cache as well for peace of mind

kiendang committed:
Remove unnecessary volume prune

We're not doing anything with volumes here

kiendang committed:
Delete redundant cleanup

kiendang commented on tox.ini:
```diff
@@ -1041,3 +1042,38 @@ commands =
+    PIP_PACKAGES = {env:PIP_PACKAGES:llama-index opendp}
+    SYSTEM_PACKAGES = {env:SYSTEM_PACKAGES:curl wget htop}
+commands =
+    bash -c 'docker rm $(docker ps -aq) --force || true'
```
@shubham3121 I actually think we should not have this cleanup here. For the test itself we're not creating any persistent container as we run everything with `docker run --rm`. Also I don't see we do anything that affects other existing containers as well.

Removing all existing containers here is a bit intrusive. If a developer has other non-PySyft related containers running this would delete them too. I know we have quite a few of these in the code base but the less we do it the better.

shubham3121 replied:
I agree. Let's only restrict to removing the containers and images we create in this job.

kiendang replied:
done

kiendang committed:
Remove unnecessary cleanup

shubham3121 commented on packages/grid/backend/backend.dockerfile:
```diff
@@ -39,15 +40,20 @@ WORKDIR $APPDIR
 
 # Install all dependencies together here to avoid any version conflicts across pkgs
-RUN --mount=type=cache,target=$HOME/.cache/,rw,uid=$UID \
+RUN --mount=type=cache,id=pip-$UID,target=$HOME/.cache/pip,uid=$UID,gid=$UID,sharing=locked \
```
👍🏾

shubham3121 commented on tox.ini:
```diff
@@ -1041,3 +1042,37 @@ commands =
+    bash -c 'for pkg in $SYSTEM_PACKAGES; do docker run --rm cpu-worker:opendp apk -e info "$pkg"; done'
+
+    bash -c '\
+        docker build \
```
🎉

shubham3121 reviewed (approved):
Great work @khoaguin and @kiendang !!! 🎉 💯

khoaguin commented:
> Great work @khoaguin and @kiendang !!! 🎉 💯

Thank you so much @shubham3121 and @kiendang for your great help on the PR!
